### PR TITLE
kubelet: Fix the volume manager didn't check the device mount state in the actual state of the world before marking the volume as detached

### LINKED ISSUE #125541: The PV may be in a Terminating state and cannot be deleted when the pod are created and then the pod and pvc are quickly force deleted .
### What happened?

In my testing environment, I discovered a strange phenomenon: when I create a pod that uses a PVC and then forcefully delete the pod and PVC before the pod creation is complete, the PV bound to the PVC **may** enter a Terminating state and cannot be deleted.

The log is as follows
```
I0614 10:04:56.153700   77153 kubelet.go:2389] "SyncLoop ADD" source="api" pods=[default/hp-volume-dbwfm]

I0614 10:04:58.233778   77153 reconciler_common.go:231] "operationExecutor.MountVolume started for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/XXXX^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") " pod="default/hp-volume-dbwfm"
I0614 10:04:58.234344   77153 operation_generator.go:1101] "MapVolume.WaitForAttach entering for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") DevicePath \"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/eac1a6c0-bdc4-4b75-804f-7f89df07694b\"" pod="default/hp-volume-dbwfm"
I0614 10:04:58.308685   77153 operation_generator.go:1111] "MapVolume.WaitForAttach succeeded for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") DevicePath \"csi-75dac338e0aff53508bd6995ba2aa58d65d6eb42310c3d9d7111f2720a1d2823\"" pod="default/hp-volume-dbwfm"

I0614 10:05:02.444088   77153 kubelet.go:2405] "SyncLoop DELETE" source="api" pods=[default/hp-volume-dbwfm]
I0614 10:05:02.520860   77153 kubelet.go:2399] "SyncLoop REMOVE" source="api" pods=[default/hp-volume-dbwfm]

I0614 10:05:17.729528   77153 operation_generator.go:1225] "MapVolume.MapPodDevice succeeded for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") volumeMapPath \"/var/lib/kubelet/pods/42077f53-c1df-4b6f-8647-f06100647095/volumeDevices/kubernetes.io~csi\"" pod="default/hp-volume-dbwfm"
I0614 10:05:17.734612   77153 operation_generator.go:1233] MapVolume.NodeExpandVolume failed for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") pod "hp-volume-dbwfm" (UID: "42077f53-c1df-4b6f-8647-f06100647095") : mountVolume.NodeExpandVolume get PVC failed : persistentvolumeclaims "pvc-volume-251e86b0-cfe7-475b-88f4-cced2416d3ff" is forbidden: User "system:node:worker3" cannot get resource "persistentvolumeclaims" in API group "" in the namespace "default": no relationship found between node 'worker3' and this object
I0614 10:05:17.734835   77153 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820 podName: nodeName:}" failed. No retries permitted until 2024-06-14 10:05:18.734806235 +0800 CST m=+62756.584763139 (durationBeforeRetry 1s). Error: MapVolume.MarkVolumeAsMounted failed while expanding volume for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") pod "hp-volume-dbwfm" (UID: "42077f53-c1df-4b6f-8647-f06100647095") : mountVolume.NodeExpandVolume get PVC failed : persistentvolumeclaims "pvc-volume-251e86b0-cfe7-475b-88f4-cced2416d3ff" is forbidden: User "system:node:worker3" cannot get resource "persistentvolumeclaims" in API group "" in the namespace "default": no relationship found between node 'worker3' and this object
I0614 10:05:18.795992   77153 operation_generator.go:1101] "MapVolume.WaitForAttach entering for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-r5sp7\" (UID: \"eac1a6c0-bdc4-4b75-804f-7f89df07694b\") DevicePath \"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/42077f53-c1df-4b6f-8647-f06100647095\"" pod="default/hp-volume-r5sp7"
I0614 10:06:03.681110   77153 reconciler_common.go:231] "operationExecutor.MountVolume started for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") " pod="default/hp-volume-dbwfm"
I0614 10:06:03.681366   77153 operation_generator.go:1101] "MapVolume.WaitForAttach entering for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") DevicePath \"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/7fb78f1b-06f4-4b95-8514-a80de1f0294d\"" pod="default/hp-volume-dbwfm"
I0614 10:06:03.726906   77153 operation_generator.go:1111] "MapVolume.WaitForAttach succeeded for volume \"pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820\" (UniqueName: \"kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820\") pod \"hp-volume-dbwfm\" (UID: \"42077f53-c1df-4b6f-8647-f06100647095\") DevicePath \"csi-75dac338e0aff53508bd6995ba2aa58d65d6eb42310c3d9d7111f2720a1d2823\"" pod="default/hp-volume-dbwfm"
I0614 10:06:15.315317   77153 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820 podName: nodeName:}" failed. No retries permitted until 2024-06-14 10:06:15.815290561 +0800 CST m=+62813.665247389 (durationBeforeRetry 500ms). Error: MapVolume.MapPodDevice failed for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") pod "hp-volume-dbwfm" (UID: "42077f53-c1df-4b6f-8647-f06100647095") : rpc error: code = Internal desc = failed to check client lun mapping status for volume pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820
I0614 10:06:16.176459   77153 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820 podName: nodeName:}" failed. No retries permitted until 2024-06-14 10:06:16.676425647 +0800 CST m=+62814.526382436 (durationBeforeRetry 500ms). Error: UnmapDevice failed for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") on node "worker3" : the device "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev" is still referenced from other Pods [/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev/42077f53-c1df-4b6f-8647-f06100647095]
I0614 10:06:16.696898   77153 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820 podName: nodeName:}" failed. No retries permitted until 2024-06-14 10:06:17.696865394 +0800 CST m=+62815.546822186 (durationBeforeRetry 1s). Error: UnmapDevice failed for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") on node "worker3" : the device "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev" is still referenced from other Pods [/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev/42077f53-c1df-4b6f-8647-f06100647095]
I0614 10:06:17.742271   77153 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820 podName: nodeName:}" failed. No retries permitted until 2024-06-14 10:06:19.742240671 +0800 CST m=+62817.592197458 (durationBeforeRetry 2s). Error: UnmapDevice failed for volume "pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820" (UniqueName: "kubernetes.io/csi/iscsi.csi.xsky.com^csi-iscsi-pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820") on node "worker3" : the device "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev" is still referenced from other Pods [/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820/dev/42077f53-c1df-4b6f-8647-f06100647095]

```

kubectl get pv pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820
```
pv-dc3407bc-6c3f-4123-bcf8-8f28b7575820   2Gi       RWX            Delete           Terminating   default/pvc-dc3407bc-6c3f-4123-bcf8-8f28b7575820  csi-XXX            1h
```

 I drawn up a time-series chart for the log contents .

image

### What did you expect to happen?

the PV can be delete successfully

### How can we reproduce it (as minimally and precisely as possible)?

1. create a pvc , the bound pv's persistentVolumeReclaimPolicy is **Delete**;
2. create a pod use the pvc;
3. force delete the pod and the pvc（kubectl delete po xx --force && kubectl delete pvc yy --force ）
4. It is possible that the PV may be in a Terminating state and cannot be deleted

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>
1.28

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### LINKED ISSUE #127520: [Flaky Test] k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler Test_UncertainDeviceGlobalMounts
### Which jobs are flaking?
* master-blocking
  * ci-kubernetes-unit

### Which tests are flaking?
`k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler`

### Since when has it been flaking?
Flaked on September 20, 2024  
[Link to failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1837283673959829504)

### Testgrid link
[TestGrid: ci-kubernetes-unit](https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit)

### Reason for failure (if possible)
```
{Failed;Failed; === RUN Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ I0921 00:32:12.806506 56545 operation_generator.go:1014] "MapVolume.WaitForAttach entering for volume "success-and-timeout-device-name"" pod="pod1" E0921 00:32:12.807049 56545 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-6ec1e204152d0da11b247a1c9de569d6: no such file or directory" --- FAIL: Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ (0.60s) }
```

### Anything else we need to know?
Failed on commit 52095a8b but also passed twice - thus flaky.

#### 1.32.0-alpha.1 is due to be cut on October 1st.
Is this an issue that could block this next release cut? Please let me know. Thank you!

### Relevant SIG(s)
/sig storage  
/cc @kubernetes/release-team-release-signal
/test flake

### LINKED ISSUE #128126: Failure cluster [160f0676...]: Error verifying UnMountDeviceCallCount: Expected DeviceUnmount Call 1, got 0
##### Error text:
```
Failed;Failed;
=== RUN   Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Filesystem][
=== PAUSE Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Filesystem][
=== CONT  Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Filesystem][
E1008 21:58:23.121079   43496 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-3e017bc9e2cc0330461d202a228a4439: no such file or directory"
I1008 21:58:23.121129   43496 reconciler.go:26] "Reconciler: start to sync state"
I1008 21:58:23.121285   43496 reconciler_common.go:251] "operationExecutor.VerifyControllerAttachedVolume started for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") " pod="pod1"
I1008 21:58:23.169125   43496 operation_generator.go:1491] "Controller attach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") device path: \"fake/path\"" pod="pod1"
I1008 21:58:23.264405   43496 operation_generator.go:538] "MountVolume.WaitForAttach entering for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"fake/path\"" pod="pod1"
I1008 21:58:23.264673   43496 operation_generator
```    reconciler_test.go:1654: Error verifying UnMountDeviceCallCount: Expected DeviceUnmount Call 1, got 0
#### Recent failures:
[10/8/2024, 11:42:09 PM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit/1843768670367518720)

/kind failing-test
/kind flake
/sig storage

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

See https://github.com/kubernetes/kubernetes/issues/127520#issuecomment-2419280422

```
(base) ➜  ~ cat /var/folders/n6/33y75ytn36g38k2wnkb6lp8r0000gn/T/go-stress-20241021T135714-609664758
E1021 13:58:06.267415   43361 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-6ec1e204152d0da11b247a1c9de569d6: no such file or directory"
I1021 13:58:06.267546   43361 reconciler.go:26] "Reconciler: start to sync state"
I1021 13:58:06.267724   43361 reconciler_common.go:251] "operationExecutor.VerifyControllerAttachedVolume started for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") " pod="pod1"
I1021 13:58:06.315859   43361 operation_generator.go:1491] "Controller attach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") device path: \"fake/path\"" pod="pod1"
I1021 13:58:06.448432   43361 operation_generator.go:1014] "MapVolume.WaitForAttach entering for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"fake/path\"" pod="pod1"
I1021 13:58:06.466757   43361 operation_generator.go:1024] "MapVolume.WaitForAttach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"/dev/sdb\"" pod="pod1"
E1021 13:58:06.501953   43361 nestedpendingoperations.go:348] Operation for "{volumeName:fake-plugin/timeout-mount-device-volume podName: nodeName:}" failed. No retries permitted until 2024-10-21 13:58:07.001894 +0800 CST m=+48.058283126 (durationBeforeRetry 500ms). Error: MapVolume.SetUpDevice failed for volume "timeout-mount-device-volume" (UniqueName: "fake-plugin/timeout-mount-device-volume") pod "pod1" (UID: "pod1uid") : mount failed
I1021 13:58:07.074835   43361 reconciler_common.go:290] "compare volume" attachedVolume={"VolumeName":"fake-plugin/timeout-mount-device-volume","VolumeSpec":{"Volume":null,"PersistentVolume":{"metadata":{"name":"timeout-mount-device-volume","uid":"pvuid","creationTimestamp":null},"spec":{"claimRef":{"name":"pvc"},"volumeMode":"Block"},"status":{}},"ReadOnly":false,"InlineVolumeSpecForCSIMigration":false,"Migrated":false},"NodeName":"mynodename","PluginIsAttachable":true,"DevicePath":"fake/path","DeviceMountPath":"","PluginName":"fake-plugin","DeviceMountState":"DeviceNotMounted","SELinuxMountContext":""} asw=[{"VolumeName":"fake-plugin/timeout-mount-device-volume","VolumeSpec":{"Volume":null,"PersistentVolume":{"metadata":{"name":"timeout-mount-device-volume","uid":"pvuid","creationTimestamp":null},"spec":{"claimRef":{"name":"pvc"},"volumeMode":"Block"},"status":{}},"ReadOnly":false,"InlineVolumeSpecForCSIMigration":false,"Migrated":false},"NodeName":"mynodename","PluginIsAttachable":true,"DevicePath":"/dev/sdb","DeviceMountPath":"plugins/fake-plugin/volumeDevices/pluginDependentPath","PluginName":"fake-plugin","DeviceMountState":"DeviceMountUncertain","SELinuxMountContext":""}]
I1021 13:58:07.074943   43361 reconciler_common.go:295] "Volume detached for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") on node \"mynodename\" DevicePath \"fake/path\""
--- FAIL: Test_UncertainDeviceGlobalMounts (0.00s)
    --- FAIL: Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ (1.64s)
        reconciler_test.go:1655: Error verifying UnMountDeviceCallCount: No Unmapper have expected TearDownDeviceCallCount. Expected: <1>.
```
I added a log for debugging purposes, see `I1021 13:58:07.074835   43361 reconciler_common.go:290` from the above test. the  DeviceMountState is `DeviceNotMounted` but the actual state is `DeviceMountUncertain`. if the state has been changed, we can not do detach action.

With this patch: no flaky within 15m

```
(base) ➜  kubernetes git:(master) ✗ stress -p 300 ./reconciler.test -test.run "^Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_\[Block\]\[$"
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
20s: 0 runs so far, 0 failures
...
14m40s: 3634 runs so far, 0 failures
14m45s: 3653 runs so far, 0 failures
14m50s: 3672 runs so far, 0 failures
14m55s: 3690 runs so far, 0 failures
15m0s: 3702 runs so far, 0 failures
^C
```

#### Which issue(s) this PR fixes:

Fixes #127520
Fixes #128126

#### Special notes for your reviewer:

We need to backport it.

Ref #125541 #123960 https://github.com/kubernetes/kubernetes/issues/113289

#### Does this PR introduce a user-facing change?

```release-note
kubelet: Fix the volume manager didn't check the device mount state in the actual state of the world before marking the volume as detached. It may cause a pod to be stuck in the Terminating state due to the above issue when it was deleted.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by carlory** [comment]
/cc @drewhagen @pacoxu @jsafrane 

We need to backport it once it is merged.

**Comment by gnufied** [comment]
So - we are thinking that, `MountVolume` and `unmountDetachDevices` are racing with each other. `MountVolume` has not quite finished marking device as uncertain at the time, we are checking if device should be unmounted and reconciler concludes that device is not mounted and hence can be marked as detached without unmounting first.

But while this is happening parallely - `MountVolume` marks device as mounted in-uncertain state, but reconciler does not see this state. But this generally means that - there must not be an operation pending for this volume. So if the test is indeed failing because of race, then operation pending check is not sufficient? Is that the conclusion here @carlory ?

**Comment by carlory** [comment]
> then operation pending check is not sufficient? 

Yes, it is. the attachedVolume may be obtained during the operation is pending, and the volume in asw may be changed after the operation is completed. At this time, when we do check, the operation is not pending, we use the old attachedVolume to decide some behaviors, which may cause some problems like above flaky tests. Re-getting the attachedVolume after the operation is completed can avoid this problem but may make the code more complex. Let's handle this in the next sync loop, it is more simple and clear. @gnufied

**Comment by carlory** [comment]
>  After checking !IsOperationPending(), can MountVolume start and finish?

@jsafrane It won't start a new MountVolume operation during the check. It is possible that there is an in-flight MountVolume operation that will mark the device as uncertain once it finishes.

Let us consider the following sequence:

reconciler's Loop 1:

1. Volume is added into asw and the device is marked as unmounted after the reconciler does the `VerifyControllerAttachedVolume`
2. the reconciler starts the MountVolume operation via the operationExecutor

reconciler's Loop 2:

1. the reconciler enters the unmountDetachDevices function and fetches all unmounted volumes from asw.
2. At this time, the in-flight MountVolume operation from Loop 1 finishes and marks the device as uncertain.
3. the unit test observes the device as uncertain and removes the pod from dsw.
4. the unmountDetachDevices function iterates over the fetched volumes from step 1, but the volume in asw is updated to uncertain. when the function does the following check `!rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName, attachedVolume.SELinuxMountContext) && !rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName, nestedpendingoperations.EmptyNodeName)`, it will pass but the local variable `attachedVolume` is not updated to reflect the new state of the volume in asw. This will cause the reconciler to mark the volume as detached before tearing down the device.

In conclusion, this PR fixes the root cause.

**Comment by carlory** [comment]
> The underlying land mine is - we need to read state from ASWO/DSOW and currently running operation's status within same critical section. We are reading them separately via two different locks. They will always be racy. :(

We re-get the device mount state from asw and compare the state with the previous fetched state only when no operation is pending for the volume. If no pending operation, we can make sure that the re-get volume state is the latest. Why it is always racy?

**Comment by carlory** [comment]
When the reconcile calls the `unmountDetachDevices` function, the `operationExecutor` can not put a new operation to the `NestedPendingOperations`. For a volume, only one operation can be in the `NestedPendingOperations` at the same time because the key is unique. Once the operation is done, it won't be executed again unless the caller puts the operation again with the same key. When the reconcile calls the `unmountDetachDevices` function, I think we don't need to worry about the operation being restarted. However, we have to handle the in-flight operation after it is completed because it may change the status of the volume.

**Comment by carlory** [comment]
> After checking !IsOperationPending(), can MountVolume start and finish?

If no new caller puts a new operation, the answer is no. `!IsOperationPending()` means no operation is pending or the operation has already been finished. 

We can see the the initial state of the operation is `operationPending: true` in the `Run` method until the operation is finished.

```go
func (grm *nestedPendingOperations) Run(
	volumeName v1.UniqueVolumeName,
	podName volumetypes.UniquePodName,
	nodeName types.NodeName,
	generatedOperations volumetypes.GeneratedOperations) error {
	grm.lock.Lock()
	defer grm.lock.Unlock()

	opKey := operationKey{volumeName, podName, nodeName}

	opExists, previousOpIndex := grm.isOperationExists(opKey)
	if opExists {
		previousOp := grm.operations[previousOpIndex]
		// Operation already exists
		if previousOp.operationPending {
			// Operation is pending
			return NewAlreadyExistsError(opKey)
		}

		backOffErr := previousOp.expBackoff.SafeToRetry(fmt.Sprintf("%+v", opKey))
		if backOffErr != nil {
			if previousOp.operationName == generatedOperations.OperationName {
				return backOffErr
			}
			// previous operation and new operation are different. reset op. name and exp. backoff
			grm.operations[previousOpIndex].operationName = generatedOperations.OperationName
			grm.operations[previousOpIndex].expBackoff = exponentialbackoff.ExponentialBackoff{}
		}

		// Update existing operation to mark as pending.
		grm.operations[previousOpIndex].operationPending = true
		grm.operations[previousOpIndex].key = opKey
	} else {
		// Create a new operation
		grm.operations = append(grm.operations,
			operation{
				key:              opKey,
				operationPending: true,
				operationName:    generatedOperations.OperationName,
				expBackoff:       exponentialbackoff.ExponentialBackoff{},
			})
	}

	go func() (eventErr, detailedErr error) {
		// Handle unhandled panics (very unlikely)
		defer k8sRuntime.HandleCrash()
		// Handle completion of and error, if any, from operationFunc()
		defer grm.operationComplete(opKey, &detailedErr)
		return generatedOperations.Run()
	}()

	return nil
}
```

**Comment by carlory** [comment]
> Looking at the code, unmountDeviceFunc has access to [actualStateOfWorld.IsVolumeMountedElsewhere()](https://github.com/kubernetes/kubernetes/blob/50998de605d59bc02fe3ece4696d034e92a99e6a/pkg/volume/util/operationexecutor/operation_executor.go#L209-L210) and it can check if the volume has still no pod local mounts. We do not have a pod name at that point, we need to check if the volume has any pod local mount. IMO that's what IsVolumeMountedElsewhere(volumeName, "") will do (please double check that and add an unit test for this case).

Thanks for your insight. I will continue to investigate the issue.

**Comment by carlory** [comment]
> What would be better is to check inside the "unmount device" operation. operationExecutor ensures that only a single operation runs at a time for a given volume with mutexes. Can we check inside [unmountDeviceFunc](https://github.com/kubernetes/kubernetes/blob/50998de605d59bc02fe3ece4696d034e92a99e6a/pkg/volume/util/operationexecutor/operation_generator.go#L853) if the operation is still needed?

+1

**Comment by carlory** [comment]
> If a new operation (such as MountVolume) starts after IsOperationPending() returns, the function may continue and enqueue volume unmount, while the other operation may succeed mounting it.

I think it won't happen. We check whether the operation is pending for a given volume with an empty pod name. According to the [`isOperationExists`](https://github.com/carlory/kubernetes/blob/master/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go#L240), it shouldn't have other operations on the same volume.

```
!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName, nestedpendingoperations.EmptyNodeName) &&
  rc.actualStateOfWorld.GetDeviceMountState(attachedVolume.VolumeName) == attachedVolume.DeviceMountState
```

> a new operation (such as MountVolume) 

Do you mean that a new pod with the same volume is added to DSW in the next reconcile loop? 

https://github.com/kubernetes/kubernetes/pull/127396 was merged last month, so the `mountOrAttachVolumes` will check whether the operation is pending and wait for it to complete. so we don't need to worry about it.

**Comment by jsafrane** [comment]
I see, `attachedVolume.DeviceMayBeMounted` can have old information. A check inside unmountDeviceFunc won't help in that case, because the unountDevice operation won't be even scheduled.

Ok, I think the new check makes sense.
* There are no other goroutines that could start MountVolume operation, only the reconciler does that.
* Therefore, after `IsOperationPending` returns, the reconciler can safely read from ASW and check the current state, no other operation can start / finish at that point.

We would need to capture this information in a comment. I would suggest the code to be more explicit, not just add a new condition to already complicated `if`:

```
if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName, attachedVolume.SELinuxMountContext) &&
    !rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName, nestedpendingoperations.EmptyNodeName) {

    // Re-read the actual state of the world, maybe the volume got mounted in the meantime.
    // <and some smart comment why it's safe, like:>
    // This is safe, because there is no pending operation (checked above) and no new operation
    // could start in the meantime. The only goroutine that adds new operations is this reconciler.
    attachedVolume = rc.actualStateOfWorld.GetDeviceMountState(attachedVolume.VolumeName)
    if attachedVolume.DeviceMayBeMounted() {
        ...
```

**Comment by carlory** [comment]
>     attachedVolume = rc.actualStateOfWorld.GetDeviceMountState(attachedVolume.VolumeName)

@jsafrane the above function doesn't return attachedVolume, I add a new function to asw.

**Comment by jsafrane** [comment]
/approve
giving @gnufied a chance to review. Sorry that we are so picky, we go bitten several times by races in the volume manager and any errors are hard to recover here.

**Comment by jsafrane** [comment]
/lgtm
I talked to @gnufied on slack.

**Comment by carlory** [comment]
Do we need to backport it? @jsafrane

**Comment by adavid44** [comment]
Hello :wave: 
We would have the need of this fix in k8s 1.30, or at least 1.31
Could it be possible to backport it ?
Many thanks :pray:

**Comment by liggitt** [comment]
This is marked as fixing a regression... can we update the release notes to indicate what release the regression was introduced?

**Comment by gnufied** [comment]
hmm, it is not a regression in strict sense. It was broken always.

**Comment by liggitt** [comment]
ok, then we should not have marked it as a regression, and should have clarified why it was so severe we were backporting changes in this area (we generally only backport security fixes and regression fixes)
